pkg/config/configobserver/configobserver_test.go

```diff
@@ -97,6 +180,62 @@ func Test_ConfigObserver_ChangeSupportConfig(t *testing.T) {
 	}
 }
 
+func Test_ConfigObserver_ConfigChanged(t *testing.T) {
+	ctrl := config.Controller{}
+	kube := kubeClientResponder{}
+	// Imitates New function
+	c := &Controller{
+		kubeClient:    &kube,
+		defaultConfig: ctrl,
+	}
+	c.mergeConfigLocked()
+
+	// Subscribe to config change event
+	configCh, closeFn := c.ConfigChanged()
+	if len(configCh) > 0 {
+		t.Fatalf("Config channel is not empty on start.")
+	}
+	// Setup mock for 1. config update
+	provideSecretMock(&kube, map[string]*corev1.Secret{
+		pullSecretKey: {Data: map[string][]byte{
+			".dockerconfigjson": nil,
+		}},
+		supportKey: {Data: map[string][]byte{
+			"username": []byte("test2"),
+		}},
+	})
+	// 1. config update
+	err := c.retrieveConfig(context.TODO())
+	if err != nil {
+		t.Fatalf("Unexpected error %s", err)
+	}
+	// Check if the event arrived on the channel
+	if len(configCh) != 1 {
+		t.Fatalf("Config channel has more/less then 1 event on a singal config change. len(configCh)==%d", len(configCh))
+	}
+
+	// Unsubscribe from config change
+	closeFn()
+	// Setup mock for 2. config update
+	provideSecretMock(&kube, map[string]*corev1.Secret{
+		pullSecretKey: {Data: map[string][]byte{
+			".dockerconfigjson": nil,
+		}},
+		supportKey: {Data: map[string][]byte{
+			"username": []byte("test3"),
+		}},
+	})
+	// 2. config update
+	err = c.retrieveConfig(context.TODO())
+	if err != nil {
+		t.Fatalf("Unexpected error %s", err)
+	}
+	// Check if unsubscribe worked. ie: no new event on the channel
+	if len(configCh) != 1 {
+		t.Fatalf("The closing function failed to unsubscribe from the config change event. len(configCh)==%d", len(configCh))
+	}
+}
```

Comment:
What about to do this test in similar fashion (declaring the testcases and iterating over them) as the above test? There are two cases (if I am not mistaken) and the code looks repeating.

Reply:
I wanted to test the scenario:
1. Subscribe to config changes
2. Config change happens ==> Event received
3. Unsubscribe from config changes
4. Config change happens ==> Event NOT received

And yeah I could probably test these in 2 iterations, but I personally don't like the "declaring the testcases and iterating over them" for few test cases (<4 or so), its too messy for a small gain. 

Also I feel that in this case having these steps happen on the same config.Controller is more "realistic".

Reply:
Fair points. Yeah I don't like the repeating code, but it's not that catastrophic in this case and in the end it's only test....